tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15358,10 +15358,66 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
   return;
 }
 
+// if this is the Entry FD, then try adding the target profile
+// shader attribute to the FD and carry on with validation
+void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD) {
+  const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;  
+  
+  // if there's no defined entry point, just return
+  if (EntryPointName.empty()) {
+    return;
+  }
+  
+  // There's no way we should be adding a shader attribute
+  // to a Function Decl if it doesn't even have an identifier
+  if (!FD->getIdentifier()) {
+    return;
+  }
+
+  // if this FD isn't the entry point, then we shouldn't add
+  // a shader attribute to this decl, so just return
+  if (EntryPointName != FD->getIdentifier()->getName()) {
+    return;
+  }
+
+  std::string profile = S.getLangOpts().HLSLProfile;
+  const ShaderModel *SM = hlsl::ShaderModel::GetByName(profile.c_str());
+  const llvm::StringRef fullName = ShaderModel::FullNameFromKind(SM->GetKind());
+
+  // don't add the attribute for an invalid profile, like library
+  if (fullName.empty()) {
+    return;
+  }
+
+  HLSLShaderAttr *currentShaderAttr = FD->getAttr<HLSLShaderAttr>();
+  // don't add the attribute if it already exists as an attribute on the decl  
+  if (currentShaderAttr) {
+    llvm::StringRef currentFullName = currentShaderAttr->getStage();
+    if (currentFullName == fullName) {
+      return;
```

Comment:
I've added an extra diag for that case. However, the node+compute case is necessary. What it's doing is preventing the addition of the compute shader attribute to a function that already has a "node" shader attribute attached. We don't want to add the compute attribute implicitly after the spec change, and it's unnecessary to do it before the spec change too, so I think I should leave this behavior as is.

Reply:
The diagnostic for conflicting attributes should be earlier - before this code is run, such as in `DiagnoseHLSLDecl`.  That should invalidate the `Declarator`, preventing us from even reaching this code when there are conflicting explicit attributes.

If there's an existing attribute that conflicts with the current shader target type, then we need to diagnose that here, which I've seen you've added.  However, I think a single new diagnostic for this situation would be clearer than reusing the `err_hlsl_attribute_mismatch` plus a new `note_hlsl_attribute_mismatch`.  Not only are those diagnostic identifiers misleading in name, the error output is confusing by being structured around an internal detail about how we create an implicit attribute from the target profile.

Compiling this:
```hlsl
#define ENTRY_TYPE "vertex"
[shader(ENTRY_TYPE)]
float4 main() : OUTPUT { ... }
```

With -T ps_6_0 would result in something like this:
```
test.hlsl:2:1: error: invalid shader stage attribute combination
[shader(ENTRY_TYPE)]
^
test.hlsl:2:1: note: implicit attribute from target profile is: pixel
[shader(ENTRY_TYPE)]
^
```

When it could be something more like this:
```
[shader(ENTRY_TYPE)]
^
test.hlsl:2:1: error: entry type 'pixel' from profile 'ps_6_0' conflicts with shader attribute type 'vertex' on entry function 'main'.
```

I also don't think we should be exempting the compute+node case here.  If we do, we will validate the entry point as a node shader, not as a compute shader, when your target profile is compute.  We don't want to support this scenario going forward, and though there are likely still tests that will break, can we remove or modify those tests before merging this change?

Next, I don't think this code should ever add a shader attribute if there's an existing shader attribute.  If there was a conflict, that should be a fatal error for considering this a valid entry point to even attempt to diagnose.  Otherwise, more confusing errors could be emitted, when this error needs to be fixed before we know how to interpret and diagnose the entry function properly.

Earlier I mentioned that the diagnostics `err_hlsl_attribute_mismatch` and `note_hlsl_attribute_mismatch` are misleading in name.  That's because there are many HLSL attributes that could "mismatch", but this error is very specific to the shader attribute, even though it doesn't say 'shader' in the name.  The note is specific to a mismatch between a shader attribute and the target profile, so that name is not clear either, though I don't think we need that one if we go with what I've suggested here.

I suggest these changes to the diagnostics:
- change `err_hlsl_attribute_mismatch` to `err_hlsl_conflicting_shader_attributes`
- remove `note_hlsl_attribute_mismatch`
- add a new `err_hlsl_profile_conflicts_with_shader_attribute` diagnostic.